Outgoing webhooks should catch 407 error more gracefully

edith007 opened the pull request:
## What is this PR for

This PR aims to solve the issue: https://github.com/zulip/zulip/issues/17555 

According to the current implementation, the message from the bot which triggered the 407 error message notifies the bot owner about the exception in the error message. 

I have done the following changes for this issue:
- [x] Modifies the 407 error message.
- [x] clear out the exception when 407 error message is triggered.
- [x] Add bot name when a outgoing_webhook is triggered.

timabbott commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -236,10 +236,13 @@ def notify_bot_owner(
 
-    notification_message = f"[A message]({message_url}) triggered an outgoing webhook."
+    notification_message = (
+        f"[A message]({message_url}) to your @**{bot.full_name}** triggered an outgoing webhook."
```
"To your bot ..." (I.e. add "bot " before the name) would be a good idea.

Also, I'd probably use the `@_` mention syntax -- we don't want this to trigger a notification to the bot under any circumstances.

timabbott commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -246,8 +246,15 @@ def notify_bot_owner(
         notification_message += "\n" + failure_message
     if status_code:
+        if status_code == 407:
+            notification_message += "\nAccess to the outgoing webhook's network is not allowed."
```
Maybe a message more like "The configured webhook URL is not network-accessible".  @alexmv may have an even better string, the main edit here is to make clear that the issue is how they configured its URL.

alexmv replied:
Maybe `The URL configured for the webhook is for a private or disallowed network.` ?

timabbott commented on zerver/tests/test_outgoing_webhook_system.py:
```diff
@@ -316,6 +317,45 @@ def test_pm_to_outgoing_webhook_bot(self, mock_requests_request: mock.Mock) -> N
+            last_message = self.get_last_message()
+            self.assertEqual(
+                last_message.content,
+                f"[A message]({message_url}) to your @**{bot.full_name}** triggered an outgoing webhook.\nAccess to the outgoing webhook's network is not allowed.",
```
I'd split this string across 2 lines at the newline character, if convenient; would be more readable.

edith007 commented:
Hi @timabbott @alexmv , I have made the suggested changes..,

- Added bot before bot name and also made the changes in the test accordingly
- Change the notification message for 407 status code.
- split the message content in `test_outgoing_webhook_system.py` across 2 lines.

timabbott commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -244,8 +244,17 @@ def notify_bot_owner(
+                f"\nThe webhook got a response with status code *{status_code}*."
+            )
+    if response_content and status_code != 407:
         notification_message += f"\nThe webhook got a response with status code *{status_code}*."
```
It seems like this is the same case as the one you have slightly above in the `else` clause -- what's your reasoning there?  

More generally I think you might be better off using an `else if` structure to make it more clear under what circumstances one might have multiple blocks.  E.g. probably you want an `if status_code == 407` as the top conditional here.

edith007 replied:
> It seems like this is the same case as the one you have slightly above in the else clause -- what's your reasoning there?
ah! this is really embarassing! this does the same thing as the previous else clause it should not be here...
also the above `else` clause is there because as per my understanding we wanted to have that message only when `status_code == 407` ... please correct me if I am wrong...

I have restructured the conditions as follows:
```
if status_code == 407:
    notification_message += f"...."
elif status_code:
     notification_message += f"...."
if response_content and status_code != 407:
    notification_message += f"...."
```
the last `status_code != 407` is there because it was independent of the status_code, if we dont have the extra check for 407, then it will display the payload in the message which we want to avoid...

timabbott commented:
This is better, but I ended up adding an additional commit that further cleans up / simplifies that error-handling function to be easy to read, and then merged this.  Hopefully that gives you a clearer sense of what I had in mind in my previous comments.

Thanks for your work on this @edith007!

edith007 commented:
This is actually makes the much more readable. My takeaway is to think some more and try to structure the code so that it's really easy to understand! Thanks for the lesson :)
